File: backend/app/schemas/stay.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.booking import BookingCreate


class CoGuestIn(BaseModel):
    guest_id: UUID
    purpose_of_visit: str | None = Field(default=None, max_length=200)
    company_name: str | None = Field(default=None, max_length=200)


class ForeignGuestIn(BaseModel):
    """Form C fields for a foreign national (FRRO compliance)."""

    passport_number: str = Field(min_length=3, max_length=32)
    passport_place_of_issue: str | None = Field(default=None, max_length=120)
    passport_expiry: date | None = None
    visa_number: str | None = Field(default=None, max_length=40)
    visa_type: str | None = Field(default=None, max_length=40)
    visa_place_of_issue: str | None = Field(default=None, max_length=120)
    visa_expiry: date | None = None
    place_of_birth: str | None = Field(default=None, max_length=120)
    country_of_birth: str | None = Field(default=None, max_length=120)
    nationality: str | None = Field(default=None, max_length=120)
    arrived_in_india_on: date | None = None
    arrival_place: str | None = Field(default=None, max_length=120)
    coming_from_city: str | None = Field(default=None, max_length=120)
    coming_from_country: str | None = Field(default=None, max_length=120)
    next_destination: str | None = Field(default=None, max_length=120)
    next_destination_country: str | None = Field(default=None, max_length=120)
    purpose_of_visit: str | None = Field(default=None, max_length=200)


class ForeignGuestOut(ForeignGuestIn):
    guest_id: UUID


class SettlementPreviewOut(BaseModel):
    """Bill preview — identical math to the actual checkout."""

    room_subtotal: Decimal
    gst_amount: Decimal
    charges_total: Decimal
    late_fee: Decimal
    discount: Decimal
    final_total: Decimal
    advance_paid: Decimal
    security_deposit: Decimal
    effective_paid: Decimal
    due: Decimal
    refund: Decimal


class CheckinPaymentIn(BaseModel):
    """Advance collected at the desk during check-in (only sent when the
    staff ticks "Payment collected")."""

    amount: Decimal = Field(gt=0)
    method: str = Field(pattern="^(cash|upi|card|bank_transfer|other)$")


class CheckinChargeIn(BaseModel):
    """Charge applied at check-in time (service chips, extra charges)."""

    description: str = Field(min_length=2, max_length=255)
    amount: Decimal = Field(gt=0)
    category: str = Field(
        default="other",
        pattern="^(food|restaurant|laundry|room_service|extra_bed|minibar|transport|damage|other)$",
    )


class CheckInRequest(BaseModel):
    booking_id: UUID
    checked_in_at: datetime | None = None
    expected_checkout_at: datetime | None = None
    co_guests: list[CoGuestIn] = Field(default_factory=list, max_length=20)
    purpose_of_visit: str | None = Field(default=None, max_length=200)
    company_name: str | None = Field(default=None, max_length=200)
    is_early: bool = False
    early_fee: Decimal = Field(default=Decimal("0.00"), ge=0)
    notes: str | None = Field(default=None, max_length=2000)
    # Digital acknowledgement of hotel terms (SRS §8).
    terms_acknowledged: bool = False
    # Form C data when the primary guest is a foreign national.
    foreign_guest: ForeignGuestIn | None = None
    # Atomic extras — applied inside the SAME transaction as the check-in so a
    # network failure can never leave a guest checked in with missing charges
    # or an unrecorded payment.
    charges: list[CheckinChargeIn] = Field(default_factory=list, max_length=20)
    advance_payment: CheckinPaymentIn | None = None


class CheckInOut(BaseModel):
    id: UUID
    booking_id: UUID
    booking_number: str
    checked_in_at: datetime
    expected_checkout_at: datetime | None
    is_early: bool
    early_fee: Decimal
    registration_numbers: list[str]


class BookAndCheckInRequest(BaseModel):
    """Walk-in flow: create the booking and check the guest in atomically.

    Both operations run in one request/transaction — if check-in validation
    fails, the booking creation rolls back too (no orphan bookings).
    """

    booking: BookingCreate
    checked_in_at: datetime | None = None
    co_guests: list[CoGuestIn] = Field(default_factory=list, max_length=20)
    purpose_of_visit: str | None = Field(default=None, max_length=200)
    company_name: str | None = Field(default=None, max_length=200)
    notes: str | None = Field(default=None, max_length=2000)
    terms_acknowledged: bool = False
    foreign_guest: ForeignGuestIn | None = None
    charges: list[CheckinChargeIn] = Field(default_factory=list, max_length=20)
    advance_payment: CheckinPaymentIn | None = None


class CurrentGuestOut(BaseModel):
    booking_id: UUID
    booking_number: str
    primary_guest_name: str
    primary_guest_phone_masked: str
    rooms: list[str]
    checked_in_at: datetime
    expected_checkout_at: datetime | None
    check_out_date: date
    payment_status: str
    due_amount: Decimal
    guest_count: int


class CurrentGuestsListOut(BaseModel):
    items: list[CurrentGuestOut]
    total: int


class RoomTransferRequest(BaseModel):
    booking_id: UUID
    from_room_id: UUID
    to_room_id: UUID
    reason: str | None = Field(default=None, max_length=1000)


class RoomTransferOut(BaseModel):
    id: UUID
    booking_id: UUID
    from_room_number: str
    to_room_number: str
    transferred_at: datetime
    reason: str | None


class CheckOutRequest(BaseModel):
    booking_id: UUID
    checked_out_at: datetime | None = None
    is_late: bool = False
    late_fee: Decimal = Field(default=Decimal("0.00"), ge=0)
    # Checkout with outstanding balance requires explicit authorization.
    allow_due: bool = False
    due_reason: str | None = Field(default=None, max_length=1000)


class CheckOutOut(BaseModel):
    id: UUID
    booking_id: UUID
    booking_number: str
    checked_out_at: datetime
    nights: int
    final_total: Decimal
    paid_amount: Decimal
    due_amount: Decimal
    refund_amount: Decimal
    is_late: bool
    late_fee: Decimal
    payment_due_authorized: bool


class CheckoutReversalRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=1000)

File: frontend/src/types/stay.ts
export interface GuestOut {
  id: string;
  full_name: string;
  normalized_phone: string;
  email: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  postal_code: string | null;
  gender: string | null;
  date_of_birth: string | null;
  id_proof_type: string | null;
  id_last4: string | null;
  id_verification_status: string;
  notes: string | null;
}

export interface GuestSearchResult {
  id: string;
  full_name: string;
  phone_masked: string;
  id_last4: string | null;
}

export interface GuestAutofill {
  id: string;
  full_name: string;
  phone: string;
  email: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  postal_code: string | null;
  gender: string | null;
  date_of_birth: string | null;
  id_proof_type: string | null;
  id_last4: string | null;
}

export interface BookingRoomOut {
  room_id: string;
  room_number: string;
  room_type_name: string;
  rate: string;
  is_current: boolean;
}

export interface BookingOut {
  id: string;
  booking_number: string;
  status: BookingStatus;
  payment_status: string;
  source: string;
  check_in_date: string;
  check_out_date: string;
  adults: number;
  children: number;
  room_count: number;
  discount_amount: string;
  tax_amount: string;
  total_amount: string;
  advance_amount: string;
  security_deposit: string;
  due_amount: string;
  special_requests: string | null;
  emergency_contact_name: string | null;
  emergency_contact_relation: string | null;
  emergency_contact_phone: string | null;
  vehicle_number: string | null;
  vehicle_type: string | null;
  parking_slot: string | null;
  primary_guest_id: string | null;
  primary_guest_name: string | null;
  primary_guest_phone: string | null;
  rooms: BookingRoomOut[];
  created_at: string;
  guest_type?: string | null;
  check_in_time?: string | null;
  check_out_time?: string | null;
}

/** One uploaded document of a booking guest (GET /api/v1/bookings/{id}/guests). */
export interface BookingGuestDocOut {
  id: string;
  document_type: string;
  side: string | null;
}

/** Registered guest of a booking (GET /api/v1/bookings/{id}/guests). */
export interface BookingGuestOut {
  guest_id: string;
  full_name: string;
  phone_masked: string;
  /** Full contact number (unmasked) — completed-bookings detail view. */
  phone: string | null;
  /** Guest address — completed-bookings detail view. */
  address: string | null;
  is_primary: boolean;
  registration_number: string;
  purpose_of_visit: string | null;
  company_name: string | null;
  id_proof_type: string | null;
  documents: BookingGuestDocOut[];
}

export type BookingStatus =
  | "pending"
  | "confirmed"
  | "checked_in"
  | "checked_out"
  | "cancelled"
  | "no_show";

export interface CurrentGuestOut {
  booking_id: string;
  booking_number: string;
  primary_guest_name: string;
  primary_guest_phone_masked: string;
  rooms: string[];
  checked_in_at: string;
  expected_checkout_at: string | null;
  check_in_date: string;
  check_out_date: string;
  check_in_time: string | null;
  check_out_time: string | null;
  payment_status: string;
  due_amount: string;
  guest_count: number;
}

export interface GuestCreatePayload {
  full_name: string;
  phone: string;
  email?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postal_code?: string;
  gender?: string;
  date_of_birth?: string;
  id_proof_type?: string;
  id_number?: string;
}

/** Staff-edited room rate — per night, or whole stay for day use. */
export interface RoomRateOverride {
  room_id: string;
  rate: string;
}

/** Booking payload nested inside BookAndCheckInRequest (mirrors backend BookingCreate). */
export interface BookingCreatePayload {
  primary_guest_id: string;
  room_ids: string[];
  rate_overrides?: RoomRateOverride[];
  check_in_date: string;
  check_out_date: string;
  adults: number;
  children: number;
  guest_type?: string | null;
  /** "HH:MM" */
  check_in_time?: string | null;
  /** "HH:MM" */
  check_out_time?: string | null;
  special_requests?: string | null;
  emergency_contact_name?: string | null;
  emergency_contact_relation?: string | null;
  emergency_contact_phone?: string | null;
  vehicle_number?: string | null;
  vehicle_type?: string | null;
  parking_slot?: string | null;
}

/**
 * Form C details for a foreign national (FRRO compliance).
 * Mirrors backend ForeignGuestIn — dates are YYYY-MM-DD strings.
 */
export interface ForeignGuestIn {
  passport_number: string;
  passport_place_of_issue?: string | null;
  passport_expiry?: string | null;
  visa_number?: string | null;
  visa_type?: string | null;
  visa_place_of_issue?: string | null;
  visa_expiry?: string | null;
  place_of_birth?: string | null;
  country_of_birth?: string | null;
  nationality?: string | null;
  arrived_in_india_on?: string | null;
  arrival_place?: string | null;
  coming_from_city?: string | null;
  coming_from_country?: string | null;
  next_destination?: string | null;
  next_destination_country?: string | null;
  purpose_of_visit?: string | null;
}

/** Co-guest sent with a check-in — optionally with their own Form C details. */
export interface CoGuestIn {
  guest_id: string;
  foreign_guest?: ForeignGuestIn | null;
}

/** Extra charge applied atomically inside the check-in transaction. */
export interface CheckInChargeIn {
  description: string;
  amount: string;
  category?: string;
}

/** Advance payment recorded atomically inside the check-in transaction. */
export interface CheckInAdvancePaymentIn {
  amount: string;
  method: string;
}

/** POST /api/v1/checkins/book-and-checkin — walk-in flow: book + check in atomically. */
export interface BookAndCheckInRequest {
  booking: BookingCreatePayload;
  checked_in_at?: string | null;
  co_guests: CoGuestIn[];
  purpose_of_visit?: string | null;
  company_name?: string | null;
  notes?: string | null;
  terms_acknowledged: boolean;
  foreign_guest?: ForeignGuestIn | null;
  charges?: CheckInChargeIn[];
  advance_payment?: CheckInAdvancePaymentIn | null;
}

/** POST /api/v1/checkins — existing-booking check-in (mirrors backend CheckInRequest). */
export interface CheckInRequest {
  booking_id: string;
  checked_in_at?: string | null;
  expected_checkout_at?: string | null;
  co_guests: CoGuestIn[];
  purpose_of_visit?: string | null;
  company_name?: string | null;
  is_early?: boolean;
  early_fee?: string;
  notes?: string | null;
  terms_acknowledged: boolean;
  foreign_guest?: ForeignGuestIn | null;
  charges?: CheckInChargeIn[];
  advance_payment?: CheckInAdvancePaymentIn | null;
  /** Staff correction of expected checkout time ("HH:MM"), stored on the booking. */
  check_out_time?: string | null;
}

export interface CheckInCreateOut {
  id: string;
  booking_id: string;
  booking_number: string;
  checked_in_at: string;
  registration_numbers: string[];
}

/**
 * GET /api/v1/checkouts/{booking_id}/preview — server-computed settlement.
 * All values are decimal strings. `final_total` = room_subtotal + GST-on-rooms
 * + charges_total (tax-inclusive) + late_fee − discount. `gst_amount` combines
 * room GST and charge taxes. `effective_paid` = advance_paid + security_deposit.
 * `due`/`refund` are already clamped to ≥ 0.
 */
export interface SettlementPreviewOut {
  room_subtotal: string;
  gst_amount: string;
  charges_total: string;
  late_fee: string;
  discount: string;
  final_total: string;
  advance_paid: string;
  security_deposit: string;
  effective_paid: string;
  due: string;
  refund: string;
}

export interface CheckOutOut {
  id: string;
  booking_id: string;
  booking_number: string;
  checked_out_at: string;
  nights: number;
  final_total: string;
  paid_amount: string;
  due_amount: string;
  refund_amount: string;
  is_late: boolean;
  late_fee: string;
  payment_due_authorized: boolean;
}
